scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+        ndb = np.zeros(self._dim)
+        ndb.fill(2)
+        iprint = -1
+        x = np.array(self._xbuffer, np.float64)
+        f = np.array(0.0, np.float64)
+        l = np.array(self._lower, np.float64)
+        u = np.array(self._upper, np.float64)
+        g = np.zeros(n, np.float64)
+        wa = np.zeros(2 * m * n + 5 * n + 11 * m * m + 8 * m, np.float64)
+        iwa = np.zeros(3 * n, np.int32)
+        task = np.zeros(1, 'S60')
+        csave = np.zeros(1, 'S60')
+        lsave = np.zeros(4, np.int32)
+        isave = np.zeros(44, np.int32)
+        dsave = np.zeros(29, np.float64)
+        task[:] = 'START'
+        if self._itsoftmax < 100:
+            self._itsoftmax = 100
+        elif self._itsoftmax > 1000:
+            self._itsoftmax = 1000
+        while True:
+            if iteration >= self._itsoftmax:
+                self._xbuffer = np.array(x)
+                self._fvalue = f
+                return
+            _lbfgsb.setulb(m, x, l, u, ndb,
+                           f, g, self._factr, self._pgtol, wa, iwa, task,
+                           iprint, csave, lsave, isave, dsave, 100)
+            iteration += 1
+            task_str = task.tostring()
+            if task_str.startswith(b'FG'):
+                self._xbuffer = np.array(x)
+                f = self._fobjective(self._xbuffer)
+                if self._knowreal:
+                    if f <= self._realthreshold:
+                        self._xbuffer = np.array(x)
+                        self._fvalue = f
+                        return
+                g = np.array(self._compute_gradient(x), np.float64)
+            elif task_str.startswith(b'NEW_X'):
+                pass
+            else:
+                self._fvalue = f
+                self._xbuffer = np.array(x)
+                return
+
+    def _compute_gradient(self, x):
+        """Computation of numerical derivatives for local search"""
+        # Can be done later here:
+        # Check here if the expression of the derivatives is provided
+        # For that, add an argument for jacobian in the main function
+        # and use MemoizeJac from optimize.py
+        g = np.zeros(self._dim, np.float64)
+        for i in range(self._dim):
+            x1 = np.array(x)
+            x2 = np.array(x)
+            respl = self._reps
+            respr = self._reps
+            x1[i] = x[i] + respr
+            if x1[i] > self._upper[i]:
+                x1[i] = self._upper[i]
+                respr = x1[i] - x[i]
+            x2[i] = x[i] - respl
+            if x2[i] < self._lower[i]:
+                x2[i] = self._lower[i]
+                respl = x[i] - x2[i]
+            f1 = self._fobjective(x1)
+            f2 = self._fobjective(x2)
+            g[i] = ((f1 - f2)) / (respl + respr)
+        idx = np.logical_or(np.isnan(g), np.isinf(g))
+        g[idx] = 101.0
+        return g
+
+    # @profile
+    def _yygas(self):
+        if self._usey == 1:
+            enter = True
```

Comment:
set `_sgas < 0` then the loop will run at least once and then you don't need the `enter` variable.